core/kazoo_attachments/src/gen_attachment.erl

```diff
@@ -0,0 +1,84 @@
+-module(gen_attachment).
+
+%% Setter
+-export([error_response/2]).
+%% Getters
+-export([error_code/1, error_body/1]).
+%% Helper(s)
+-export([is_error_response/1]).
+
+-include("kz_att.hrl").
+
+-type settings()         :: kz_data:connection().
+-type db_name()          :: kz_term:ne_binary().
+-type doc_id()           :: kz_term:ne_binary().
+-type att_name()         :: kz_term:ne_binary().
+-type contents()         :: kz_term:ne_binary().
+-type options()          :: kz_data:options().
+-type handler_props()    :: kz_data:connection().
+-type error_code()       :: pos_integer() | atom(). % 400, 404, 409, etc.
+-type error_body()       :: binary() | % encoded map()
+                            bitstring() | % <<"example">>
+                            atom(). % 'not_found' | 'return_id_missing' | etc.
+-opaque error_response() :: {'error', [{'error_code', error_code()} |
+                                       {'error_body', error_body()}]}.
+-type put_response()     :: {'ok', [{atom(), [{binary(), binary() | kz_json:object()}]}]} |
+                            error_response().
+-type fetch_response()   :: {'ok', iodata()} | error_response().
+
+-export_type([settings/0
+             ,db_name/0
+             ,doc_id/0
+             ,att_name/0
+             ,contents/0
+             ,options/0
+             ,handler_props/0
+             ,error_response/0
+             ,put_response/0
+             ,fetch_response/0
+             ]).
+
+%% =======================================================================================
+%% Callbacks
+%% =======================================================================================
+-callback put_attachment(settings()
+                        ,db_name()
+                        ,doc_id()
+                        ,att_name()
+                        ,contents()
+                        ,options()
+                        ) -> put_response().
```

Comment:
reuse the existing definitions, how many times can we redefine ne_binary() and map()

Reply:
We had a talk at one of our stand up sessions and we agreed to let it be.